acme/challenge.go

```diff
@@ -342,6 +354,258 @@ func dns01Validate(ctx context.Context, ch *Challenge, db DB, jwk *jose.JSONWebK
 	return nil
 }
 
+type WireChallengePayload struct {
+	// IDToken
+	IDToken string `json:"id_token,omitempty"`
+	// KeyAuth ({challenge-token}.{jwk-thumbprint})
+	KeyAuth string `json:"keyauth,omitempty"`
+	// AccessToken is the token generated by wire-server
+	AccessToken string `json:"access_token,omitempty"`
+}
+
+func wireOIDC01Validate(ctx context.Context, ch *Challenge, db DB, jwk *jose.JSONWebKey, payload []byte) error {
+	prov, ok := ProvisionerFromContext(ctx)
+	if !ok {
+		return NewErrorISE("no provisioner provided")
+	}
+
+	var wireChallengePayload WireChallengePayload
+	err := json.Unmarshal(payload, &wireChallengePayload)
+	if err != nil {
+		return storeError(ctx, db, ch, false, WrapError(ErrorRejectedIdentifierType, err,
+			"error unmarshalling Wire challenge payload"))
+	}
+
+	oidcOptions := prov.GetOptions().GetOIDCOptions()
+	idToken, err := oidcOptions.GetProvider(ctx).Verifier(oidcOptions.GetConfig()).Verify(ctx, wireChallengePayload.IDToken)
+	if err != nil {
+		return storeError(ctx, db, ch, false, WrapError(ErrorRejectedIdentifierType, err,
+			"error verifying ID token signature"))
+	}
+
+	var claims struct {
+		Name      string `json:"preferred_username,omitempty"`
+		Handle    string `json:"name"`
+		Issuer    string `json:"iss,omitempty"`
+		GivenName string `json:"given_name,omitempty"`
+		KeyAuth   string `json:"keyauth"` // TODO(hs): use this property instead of the one in the payload after https://github.com/wireapp/rusty-jwt-tools/tree/fix/keyauth is done
+	}
+	err = idToken.Claims(&claims)
+	if err != nil {
+		return storeError(ctx, db, ch, false, WrapError(ErrorRejectedIdentifierType, err,
+			"error retrieving claims from ID token"))
+	}
+
+	challengeValues, err := wire.ParseID([]byte(ch.Value))
+	if err != nil {
+		return WrapErrorISE(err, "error unmarshalling challenge data")
+	}
+
+	expectedKeyAuth, err := KeyAuthorization(ch.Token, jwk)
+	if err != nil {
+		return err
+	}
+	if expectedKeyAuth != wireChallengePayload.KeyAuth {
+		return storeError(ctx, db, ch, true, NewError(ErrorRejectedIdentifierType,
+			"keyAuthorization does not match; expected %s, but got %s", expectedKeyAuth, wireChallengePayload.KeyAuth))
+	}
+
+	if challengeValues.Name != claims.Name || challengeValues.Handle != claims.Handle {
+		return storeError(ctx, db, ch, false, NewError(ErrorRejectedIdentifierType, "OIDC claims don't match"))
+	}
+
+	// Update and store the challenge.
+	ch.Status = StatusValid
+	ch.Error = nil
+	ch.ValidatedAt = clock.Now().Format(time.RFC3339)
+
+	if err = db.UpdateChallenge(ctx, ch); err != nil {
+		return WrapErrorISE(err, "error updating challenge")
+	}
+
+	parsedIDToken, err := jwt.ParseSigned(wireChallengePayload.IDToken)
```

Comment:
```suggestion
	parsedIDToken, err := jose.ParseSigned(wireChallengePayload.IDToken)
```

Reply:
Fixed in https://github.com/smallstep/certificates/pull/1666/commits/70a2f431fa896624b6dafa375a47f454b3bde1a6